core/lib/dal/src/consensus_dal.rs

```diff
@@ -30,54 +30,119 @@ impl ConsensusDal<'_, '_> {
             return Ok(None);
         };
-        Ok(Some(zksync_protobuf::serde::deserialize(row.state)?))
+        let Some(genesis) = row.genesis else {
+            return Ok(None);
+        };
+        Ok(Some(zksync_protobuf::serde::deserialize(genesis)?))
     }
 
-    /// Sets the current BFT replica state.
-    pub async fn set_replica_state(&mut self, state: &ReplicaState) -> sqlx::Result<()> {
-        let state =
-            zksync_protobuf::serde::serialize(state, serde_json::value::Serializer).unwrap();
+    /// Attempts to update the genesis.
+    /// Fails if the storage contains a newer genesis (higher fork number).
+    /// Noop if the new genesis is the same as the current one.
+    /// Resets the stored consensus state otherwise and purges all certificates.
+    pub async fn try_update_genesis(&mut self, genesis: &validator::Genesis) -> anyhow::Result<()> {
+        let mut txn = self.storage.start_transaction().await?;
+        if let Some(got) = txn.consensus_dal().genesis().await? {
+            // Exit if the genesis didn't change.
+            if &got == genesis {
+                return Ok(());
+            }
+            if got.fork.number >= genesis.fork.number {
+                anyhow::bail!("transition to a past fork is not allowed");
+            }
+        }
+        let genesis =
+            zksync_protobuf::serde::serialize(genesis, serde_json::value::Serializer).unwrap();
+        let state = zksync_protobuf::serde::serialize(
+            &ReplicaState::default(),
+            serde_json::value::Serializer,
+        )
+        .unwrap();
+        sqlx::query!(
+            r#"
+            DELETE FROM miniblocks_consensus
+            "#
+        )
+        .execute(txn.conn())
+        .await?;
+        sqlx::query!(
+            r#"
+            DELETE FROM consensus_replica_state
+            "#
+        )
+        .execute(txn.conn())
+        .await?;
         sqlx::query!(
             r#"
             INSERT INTO
-                consensus_replica_state (fake_key, state)
+                consensus_replica_state (fake_key, genesis, state)
             VALUES
-                (TRUE, $1)
-            ON CONFLICT (fake_key) DO
-            UPDATE
-            SET
-                state = excluded.state
+                (TRUE, $1, $2)
             "#,
-            state
+            genesis,
+            state,
         )
-        .execute(self.storage.conn())
+        .execute(txn.conn())
         .await?;
+        txn.commit().await?;
         Ok(())
     }
 
-    /// Fetches the first consensus certificate.
-    /// Note that we didn't backfill the certificates for the past miniblocks
-    /// when enabling consensus certificate generation, so it might NOT be the certificate
-    /// for the genesis miniblock.
-    pub async fn first_certificate(&mut self) -> anyhow::Result<Option<validator::CommitQC>> {
-        let Some(row) = sqlx::query!(
+    pub async fn fork(&mut self) -> anyhow::Result<()> {
```

Comment:
A doc comment on this method would be nice, as I'd say it's not straightforward.

Separately I'll mention that it differs from how we normally use DAL: usually it's a pretty stupid wrapper around SQL that exists to prevent spreading `sqlx` all over the place, and the actual logic lives elsewhere. I don't really have an opinion on whether it's good or bad, but it also adds to the fact that the documentation is needed here.

Reply:
added documentation